File: packages/core-py/src/context_shunt/aggregate.py
"""Bounded deterministic count/distinct/grouping for validated JSON snapshots."""

from __future__ import annotations

import json
from typing import Any

from .errors import ShuntError
from .inspect import Extraction, Segment, escaped_json_cost
from .limits import Limits
from .snapshot import Snapshot, canonical_json, json_depth_and_nodes, resolve_pointer

_MAX_RETURNED_VALUES = 200
_MAX_RETURNED_GROUPS = 200
_MAX_SCALAR_BYTES = 512
_MISSING = object()
_MISSING_OUTPUT = {"missing": True}


def _optional_pointer(value: Any, pointer: str) -> Any:
    try:
        return resolve_pointer(value, pointer)
    except ShuntError as exc:
        if exc.detail == "POINTER_NOT_FOUND":
            return _MISSING
        raise


def _scalar(value: Any) -> Any:
    if value is _MISSING or value is None or isinstance(value, (str, bool)):
        return value
    raise ShuntError("INVALID_REQUEST", "BAD_SELECTOR", retryable=False)


def _output_scalar(value: Any) -> bool:
    return len(canonical_json(value).encode("utf-8")) <= _MAX_SCALAR_BYTES


def aggregate_snapshot(
    snapshot: Snapshot,
    selector: dict[str, Any],
    *,
    max_result_bytes: int,
    max_wire_bytes: int,
    max_records: int,
    limits: Limits,
) -> Extraction:
    """Aggregate in one bounded pass, including Loki-style nested JSON log records."""
    root = snapshot.json_value
    if root is None:
        if snapshot.media_type != "text/plain":
            raise ShuntError("INVALID_REQUEST", "BAD_SELECTOR", retryable=False)
        try:
            root = json.loads(snapshot.data.decode("utf-8"))
        except (UnicodeDecodeError, ValueError, RecursionError):
            raise ShuntError("INVALID_REQUEST", "BAD_JSON", retryable=False) from None
        json_depth_and_nodes(root, limits)
    outer = resolve_pointer(root, selector["records_pointer"])
    if not isinstance(outer, list):
        raise ShuntError("INVALID_REQUEST", "BAD_SELECTOR", retryable=False)

    records: list[Any] = []
    work_units = 0
    for item in outer:
        work_units += 1
        if work_units > max_records:
            raise ShuntError("LIMIT_EXCEEDED", "RESULT_OVER_SOURCE_CAP", retryable=False)
        expanded = (
            [item]
            if "expand_pointer" not in selector
            else resolve_pointer(item, selector["expand_pointer"])
        )
        if not isinstance(expanded, list):
            raise ShuntError("INVALID_REQUEST", "BAD_SELECTOR", retryable=False)
        if "expand_pointer" in selector:
            work_units += len(expanded)
        if work_units > max_records:
            raise ShuntError("LIMIT_EXCEEDED", "RESULT_OVER_SOURCE_CAP", retryable=False)
        records.extend(expanded)

    distinct_paths = selector.get("distinct", [])
    group_paths = selector.get("group_by", [])
    distinct: dict[str, dict[str, Any]] = {path: {} for path in distinct_paths}
    groups: dict[str, dict[str, Any]] = {}
    matched = 0
    parsed_bytes = 0
    parsed_nodes = 0
    filter_spec = selector.get("filter")
    filter_expected_key: str | None = None
    if filter_spec is not None and "equals" in filter_spec:
        expected = _scalar(filter_spec["equals"])
        if not _output_scalar(expected):
            raise ShuntError("INVALID_REQUEST", "BAD_SELECTOR", retryable=False)
        filter_expected_key = canonical_json(expected)

    for raw in records:
        record = (
            raw
            if "record_pointer" not in selector
            else resolve_pointer(raw, selector["record_pointer"])
        )
        if selector.get("parse_json"):
            if not isinstance(record, str):
                raise ShuntError("INVALID_REQUEST", "BAD_SELECTOR", retryable=False)
            parsed_bytes += len(record.encode("utf-8"))
            if parsed_bytes > limits.max_source_bytes:
                raise ShuntError("LIMIT_EXCEEDED", "RESULT_OVER_SOURCE_CAP", retryable=False)
            try:
                record = json.loads(record)
            except (ValueError, RecursionError):
                raise ShuntError("INVALID_REQUEST", "BAD_JSON", retryable=False) from None
            _, nodes = json_depth_and_nodes(record, limits)
            parsed_nodes += nodes
            if parsed_nodes > limits.json_max_nodes:
                raise ShuntError("LIMIT_EXCEEDED", "JSON_TOO_MANY_NODES", retryable=False)

        if filter_spec is not None:
            candidate = _optional_pointer(record, filter_spec["pointer"])
            if candidate is _MISSING:
                continue
            if "equals" in filter_spec:
                if not (candidate is None or isinstance(candidate, (str, bool))):
                    continue
                candidate = _scalar(candidate)
                if (
                    not _output_scalar(candidate)
                    or canonical_json(candidate) != filter_expected_key
                ):
                    continue
            elif not isinstance(candidate, str) or filter_spec["contains"] not in candidate:
                continue
        matched += 1

        for path in distinct_paths:
            value = _scalar(_optional_pointer(record, path))
            if value is not _MISSING:
                distinct[path][canonical_json(value)] = value
        if group_paths:
            key = []
            for path in group_paths:
                value = _scalar(_optional_pointer(record, path))
                key.append(_MISSING_OUTPUT if value is _MISSING else value)
            encoded = canonical_json(key)
            if encoded in groups:
                groups[encoded]["count"] += 1
            else:
                groups[encoded] = {"key": key, "count": 1}

    distinct_rows: list[dict[str, Any]] = []
    for path in distinct_paths:
        all_values = sorted(distinct[path].items(), key=lambda item: item[0].encode("utf-8"))
        values = [value for _, value in all_values if _output_scalar(value)][:_MAX_RETURNED_VALUES]
        distinct_rows.append(
            {
                "path": path,
                "count": len(all_values),
                "values": values,
                "values_complete": len(values) == len(all_values),
            }
        )
    all_groups = sorted(groups.items(), key=lambda item: item[0].encode("utf-8"))
    group_rows = [
        row for _, row in all_groups if all(_output_scalar(value) for value in row["key"])
    ][:_MAX_RETURNED_GROUPS]
    groups_complete = len(group_rows) == len(all_groups)

    def build() -> str:
        return canonical_json(
            {
                "distinct": distinct_rows,
                "group_by": group_paths,
                "group_count": len(all_groups),
                "groups": group_rows,
                "groups_complete": groups_complete,
                "matched_count": matched,
                "records_scanned": len(records),
                "schema": "context_shunt.aggregate.v1",
            }
        )

    text = build()

    def fits() -> bool:
        return len(text.encode("utf-8")) <= max_result_bytes and (
            Segment.wire_overhead("aggregate", 0, len(records)) + escaped_json_cost(text)
            <= max_wire_bytes
        )

    while not fits() and group_rows:
        group_rows.pop()
        groups_complete = False
        text = build()
    while not fits() and any(row["values"] for row in distinct_rows):
        row = next(row for row in reversed(distinct_rows) if row["values"])
        row["values"].pop()
        row["values_complete"] = False
        text = build()
    if not fits():
        raise ShuntError("LIMIT_EXCEEDED", "UNIT_OVER_PAGE_BUDGET", retryable=False)

    return Extraction(
        mode="aggregate",
        segments=[Segment(kind="aggregate", start=0, end=len(records), text=text)],
        result_bytes=len(text.encode("utf-8")),
        complete=True,
        # Aggregate traverses parsed records, not the text line index.
        lines_scanned=0,
        records_scanned=len(records),
        records_matched=matched,
    )


__all__ = ["aggregate_snapshot"]

File: packages/core-ts/src/aggregate.ts
/** Bounded deterministic aggregation over an already-validated JSON snapshot. */
import { ShuntError } from "./errors.js";
import { type Extraction, escapedJsonCost, segmentWireOverhead } from "./inspect.js";
import { type Limits } from "./limits.js";
import { canonicalJson, jsonDepthAndNodes, resolvePointer, type Snapshot } from "./snapshot.js";

const MAX_RETURNED_VALUES = 200;
const MAX_RETURNED_GROUPS = 200;
const MAX_SCALAR_BYTES = 512;
const MISSING = Symbol("missing");
const MISSING_OUTPUT = Object.freeze({ missing: true as const });
type Scalar = string | number | boolean | null;
type GroupValue = Scalar | typeof MISSING_OUTPUT;

function optionalPointer(value: unknown, pointer: string): unknown | typeof MISSING {
  try {
    return resolvePointer(value, pointer);
  } catch (err) {
    if (err instanceof ShuntError && err.detail === "POINTER_NOT_FOUND") return MISSING;
    throw err;
  }
}

function scalar(value: unknown): Scalar | typeof MISSING {
  if (value === MISSING) return MISSING;
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
    }
    return value;
  }
  if (value === null || ["string", "boolean"].includes(typeof value)) {
    return value as string | boolean | null;
  }
  throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
}

function scalarKey(value: Scalar): string {
  return canonicalJson(value);
}

function canonicalOrder(left: string, right: string): number {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}

function outputScalar(value: GroupValue): boolean {
  return Buffer.byteLength(canonicalJson(value), "utf8") <= MAX_SCALAR_BYTES;
}

export interface AggregateSelector {
  records_pointer: string;
  expand_pointer?: string;
  record_pointer?: string;
  parse_json?: boolean;
  filter?: { pointer: string; equals?: unknown; contains?: string };
  group_by?: string[];
  distinct?: string[];
}

/**
 * Aggregate every selected record in one pass. The source JSON has already passed the
 * snapshot depth/node caps; embedded JSON strings are independently checked and share the
 * same cumulative node and byte ceilings, so minified Loki payloads cannot amplify work.
 */
export function aggregateSnapshot(
  snapshot: Snapshot,
  selector: AggregateSelector,
  opts: { maxResultBytes: number; maxWireBytes: number; maxRecords: number; limits: Limits },
): Extraction {
  let root = snapshot.jsonValue;
  if (root === undefined) {
    if (snapshot.mediaType !== "text/plain") {
      throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
    }
    try {
      root = JSON.parse(Buffer.from(snapshot.data).toString("utf8"));
    } catch {
      throw new ShuntError("INVALID_REQUEST", "BAD_JSON", false);
    }
    jsonDepthAndNodes(root, opts.limits);
  }
  const outer = resolvePointer(root, selector.records_pointer);
  if (!Array.isArray(outer)) {
    throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
  }

  const records: unknown[] = [];
  let workUnits = 0;
  for (const item of outer) {
    workUnits += 1;
    if (workUnits > opts.maxRecords) {
      throw new ShuntError("LIMIT_EXCEEDED", "RESULT_OVER_SOURCE_CAP", false);
    }
    const expanded = selector.expand_pointer === undefined
      ? [item]
      : resolvePointer(item, selector.expand_pointer);
    if (!Array.isArray(expanded)) {
      throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
    }
    if (selector.expand_pointer !== undefined) workUnits += expanded.length;
    if (workUnits > opts.maxRecords) {
      throw new ShuntError("LIMIT_EXCEEDED", "RESULT_OVER_SOURCE_CAP", false);
    }
    records.push(...expanded);
  }

  const distinctPaths = selector.distinct ?? [];
  const groupPaths = selector.group_by ?? [];
  const distinct = new Map<string, Map<string, Scalar>>(
    distinctPaths.map((path) => [path, new Map()]),
  );
  const groups = new Map<string, { key: GroupValue[]; count: number }>();
  let matched = 0;
  let parsedBytes = 0;
  let parsedNodes = 0;
  const filter = selector.filter;
  let filterExpectedKey: string | undefined;
  if (filter !== undefined && Object.prototype.hasOwnProperty.call(filter, "equals")) {
    const expected = scalar(filter.equals);
    if (expected === MISSING || !outputScalar(expected)) {
      throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
    }
    filterExpectedKey = canonicalJson(expected);
  }

  for (const raw of records) {
    let record = selector.record_pointer === undefined
      ? raw
      : resolvePointer(raw, selector.record_pointer);
    if (selector.parse_json) {
      if (typeof record !== "string") {
        throw new ShuntError("INVALID_REQUEST", "BAD_SELECTOR", false);
      }
      parsedBytes += Buffer.byteLength(record, "utf8");
      if (parsedBytes > opts.limits.maxSourceBytes) {
        throw new ShuntError("LIMIT_EXCEEDED", "RESULT_OVER_SOURCE_CAP", false);
      }
      try {
        record = JSON.parse(record);
      } catch {
        throw new ShuntError("INVALID_REQUEST", "BAD_JSON", false);
      }
      parsedNodes += jsonDepthAndNodes(record, opts.limits).nodes;
      if (parsedNodes > opts.limits.jsonMaxNodes) {
        throw new ShuntError("LIMIT_EXCEEDED", "JSON_TOO_MANY_NODES", false);
      }
    }

    if (filter !== undefined) {
      const candidate = optionalPointer(record, filter.pointer);
      if (candidate === MISSING) continue;
      if (Object.prototype.hasOwnProperty.call(filter, "equals")) {
        if (candidate !== null && !["string", "number", "boolean"].includes(typeof candidate)) {
          continue;
        }
        const normalizedCandidate = scalar(candidate);
        if (normalizedCandidate === MISSING || !outputScalar(normalizedCandidate)
          || canonicalJson(normalizedCandidate) !== filterExpectedKey) continue;
      } else if (typeof candidate !== "string" || !candidate.includes(filter.contains ?? "")) {
        continue;
      }
    }
    matched += 1;

    for (const path of distinctPaths) {
      const value = scalar(optionalPointer(record, path));
      if (value !== MISSING) distinct.get(path)?.set(scalarKey(value), value);
    }
    if (groupPaths.length > 0) {
      const key = groupPaths.map((path) => {
        const value = scalar(optionalPointer(record, path));
        return value === MISSING ? MISSING_OUTPUT : value;
      });
      const encoded = canonicalJson(key);
      const prior = groups.get(encoded);
      if (prior) prior.count += 1;
      else groups.set(encoded, { key, count: 1 });
    }
  }

  const distinctRows = distinctPaths.map((path) => {
    const all = [...(distinct.get(path)?.entries() ?? [])].sort(([a], [b]) => canonicalOrder(a, b));
    const values = all.filter(([, value]) => outputScalar(value))
      .slice(0, MAX_RETURNED_VALUES).map(([, value]) => value);
    return { path, count: all.length, values, values_complete: values.length === all.length };
  });
  const allGroups = [...groups.entries()].sort(([a], [b]) => canonicalOrder(a, b));
  let groupRows = allGroups.filter(([, row]) => row.key.every(outputScalar))
    .slice(0, MAX_RETURNED_GROUPS).map(([, row]) => row);
  let groupsComplete = groupRows.length === allGroups.length;

  const build = (): string => canonicalJson({
    distinct: distinctRows,
    group_by: groupPaths,
    group_count: allGroups.length,
    groups: groupRows,
    groups_complete: groupsComplete,
    matched_count: matched,
    records_scanned: records.length,
    schema: "context_shunt.aggregate.v1",
  });
  let text = build();
  const fits = (): boolean => {
    const bytes = Buffer.byteLength(text, "utf8");
    return bytes <= opts.maxResultBytes
      && segmentWireOverhead("aggregate", 0, records.length) + escapedJsonCost(text)
        <= opts.maxWireBytes;
  };
  while (!fits() && groupRows.length > 0) {
    groupRows = groupRows.slice(0, -1);
    groupsComplete = false;
    text = build();
  }
  while (!fits() && distinctRows.some((row) => row.values.length > 0)) {
    const row = [...distinctRows].reverse().find((candidate) => candidate.values.length > 0)!;
    row.values = row.values.slice(0, -1);
    row.values_complete = false;
    text = build();
  }
  if (!fits()) throw new ShuntError("LIMIT_EXCEEDED", "UNIT_OVER_PAGE_BUDGET", false);
  const resultBytes = Buffer.byteLength(text, "utf8");
  return {
    mode: "aggregate",
    segments: [{ kind: "aggregate", start: 0, end: records.length, text }],
    resultBytes,
    complete: true,
    nextCursorState: undefined,
    // Aggregate traverses parsed records, not the text line index.
    linesScanned: 0,
    recordsScanned: records.length,
    recordsMatched: matched,
    scanBudgetExhausted: false,
    matchesFound: undefined,
    stalled: false,
    stallReason: "content",
  };
}
